[Unified PDF] Active form entry fields can get detached from the document on zoom out

pxlcoder opened the pull request:
#### 4e5ea69a3c5ff8a548af422b9df424616dd53c97
<pre>
[Unified PDF] Active form entry fields can get detached from the document on zoom out
<a href="https://bugs.webkit.org/show_bug.cgi?id=272596">https://bugs.webkit.org/show_bug.cgi?id=272596</a>
<a href="https://rdar.apple.com/126222406">rdar://126222406</a>

Reviewed by Abrar Rahman Protyasha.

Annotation geometry currently gets updated when zooming in/out as a consequence
of the scroll position changing, via `UnifiedPDFPlugin::didChangeScrollOffset`.

However, when zooming out far enough, the scroll position stops changing, and
there is nothing triggering geometry updates for annotations.

Fix by ensuring annotation geometry is up-to-date whenever the scale factor changes.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::setScaleFactor):

Canonical link: <a href="https://commits.webkit.org/277438@main">https://commits.webkit.org/277438@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9e29ef550a6fbc488ea268b63209c0bceff860bc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47603 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26790 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50274 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50285 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43652 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32564 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24252 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38759 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48185 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24402 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41013 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20058 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21883 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5645 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43934 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42617 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52169 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22637 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18967 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46065 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23910 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45100 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24699 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6724 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23629 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1130,6 +1130,9 @@ static String mutationObserverNotificationString()
     updatePageBackgroundLayers();
     updateSnapOffsets();
 
+    if (m_activeAnnotation)
+        m_activeAnnotation->updateGeometry();
```
You need to wrap it in `#if PLATFORM(MAC)`

aprotyas reviewed (commented):
When we zoom in, so both the scale factor and scroll offset change, does this behave correctly and is just a duplication of work? (Thus benign, but could be optimized)

pxlcoder committed:
[Unified PDF] Active form entry fields can get detached from the document on zoom out
https://bugs.webkit.org/show_bug.cgi?id=272596
rdar://126222406

Reviewed by Abrar Rahman Protyasha.

Annotation geometry currently gets updated when zooming in/out as a consequence
of the scroll position changing, via `UnifiedPDFPlugin::didChangeScrollOffset`.

However, when zooming out far enough, the scroll position stops changing, and
there is nothing triggering geometry updates for annotations.

Fix by ensuring annotation geometry is up-to-date whenever the scale factor changes.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::setScaleFactor):

Canonical link: https://commits.webkit.org/277438@main
